sql-plugin/src/main/scala/com/nvidia/spark/rapids/io/async/TrafficController.scala

```diff
@@ -0,0 +1,142 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.nvidia.spark.rapids.io.async
+
+import java.util.concurrent.Callable
+import javax.annotation.concurrent.GuardedBy
+
+import com.nvidia.spark.rapids.RapidsConf
+
+/**
+ * Simple wrapper around a [[Callable]] that also keeps track of the host memory bytes used by
+ * the task.
+ *
+ * Note: we may want to add more metadata to the task in the future, such as the device memory,
+ * as we implement more throttling strategies.
+ */
+class Task[T](val hostMemoryBytes: Long, callable: Callable[T]) extends Callable[T] {
+  override def call(): T = callable.call()
+}
+
+/**
+ * Throttle interface to be implemented by different throttling strategies.
+ *
+ * Currently, only HostMemoryThrottle is implemented, which limits the maximum in-flight host
+ * memory bytes. In the future, we can add more throttling strategies, such as limiting the
+ * device memory usage, the number of tasks, etc.
+ */
+trait Throttle {
+
+  /**
+   * Returns true if the task can be accepted, false otherwise.
+   * TrafficController will block the task from being scheduled until this method returns true.
+   */
+  def canAccept[T](task: Task[T]): Boolean
+
+  /**
+   * Callback to be called when a task is scheduled.
+   */
+  def taskScheduled[T](task: Task[T]): Unit
+
+  /**
+   * Callback to be called when a task is completed, either successfully or with an exception.
+   */
+  def taskCompleted[T](task: Task[T]): Unit
+}
+
+/**
+ * Throttle implementation that limits the total host memory used by the in-flight tasks.
+ */
+class HostMemoryThrottle(val maxInFlightHostMemoryBytes: Long) extends Throttle {
+  private var totalHostMemoryBytes: Long = 0
+
+  override def canAccept[T](task: Task[T]): Boolean = {
+    totalHostMemoryBytes + task.hostMemoryBytes <= maxInFlightHostMemoryBytes
+  }
+
+  override def taskScheduled[T](task: Task[T]): Unit = {
+    totalHostMemoryBytes += task.hostMemoryBytes
+  }
+
+  override def taskCompleted[T](task: Task[T]): Unit = {
+    totalHostMemoryBytes -= task.hostMemoryBytes
+  }
+
+  def getTotalHostMemoryBytes: Long = totalHostMemoryBytes
+}
+
+/**
+ * TrafficController is responsible for blocking tasks from being scheduled when the throttle
+ * is exceeded. It also keeps track of the number of tasks that are currently scheduled.
+ *
+ * This class is thread-safe as it is used by multiple tasks.
+ */
+class TrafficController protected[rapids] (throttle: Throttle) {
+
+  @GuardedBy("this")
+  private var numTasks: Int = 0
+
+  /**
+   * Blocks the task from being scheduled until the throttle allows it. If there is no task
+   * currently scheduled, the task is scheduled immediately even if the throttle is exceeded.
+   */
+  def blockUntilRunnable[T](task: Task[T]): Unit = synchronized {
+    if (numTasks > 0) {
+      while (!throttle.canAccept(task)) {
```

Comment:
it might be cleaner, if `Throttle` has just one method here. Thoughts?

```
while (!throttle.taskScheduled(task)) {
  wait(100)
}
// task scheduled
numTasks += 1
```

Reply:
Also I feel that throttle should handle the case where this is the first task, so I don't think we need `numTasks`. These are nits by the way.

Reply:
As for the making the interface cleaner, I see your point but am not sure if that's cleaner. If we do so, `taskScheduled()` (or we should probably call it `trySchedule()`) would have a side effect of updating `totalHostMemoryBytes` besides checking the eligibility of the task running. I could argue that this is less cleaner because the interface enforces a side effect. I'd like to leave this as it is for now.

As for tracking `numTasks` in the throttle, I intended to have each throttle do one job and make them composable. For example, we could add a new `DeviceMemoryThrottle` that limits in-flight device memory bytes. To use this and `HostMemoryThrottle` together, we could also add a new `CompositeThrottle` that would look like this:

```scala
class CompositeThrottle(throttles: Throttle*) extends Throttle {
  override def canAccept[T](task: Task[T]): Boolean = {
    throttles.forall(_.canAccept(task))
  }

  override def taskScheduled[T](task: Task[T]): Unit = {
    throttles.foreach(_.taskScheduled(task))
  }

  override def taskCompleted[T](task: Task[T]): Unit = {
    throttles.foreach(_.taskCompleted(task))
  }
}
```

The idea behind this design is that composable throttle will make each implementation simple and easy to add/remove/modify them. However, this design will require every throttle to track `numTasks` if we put it the throttle. Also, `numTasks` is special as the controller should accept any task if `numTasks` is 0. It's not a throttle, but a bypass to work around the throttle.

Reply:
That sounds good to me. I like the composeable example.